### PR TITLE
Core respawn priority fix

### PR DESCRIPTION
**Pull request by Slava0135**
https://user-images.githubusercontent.com/40753025/104125912-c879db00-536a-11eb-80cf-4ed7ae565f79.mp4

### CODE REVIEW

**Comment by Slava0135** [comment]
So basically, you respawn in the closest biggest core and not just the closest.

File: core/src/mindustry/entities/comp/PlayerComp.java [hunk 1]
```diff
@@ -55,6 +55,18 @@ public boolean isBuilder(){
         return state.teams.closestCore(x, y, team);
     }
 
+    public @Nullable CoreBuild closestBestCore(){
```
Comment (Quezler) [inline core/src/mindustry/entities/comp/PlayerComp.java:58]:
```suggestion
    public @Nullable CoreBuild respawnCore(){
```

File: core/src/mindustry/entities/comp/PlayerComp.java [hunk 2]
```diff
@@ -55,6 +55,18 @@ public boolean isBuilder(){
...
+        CoreBuild result = null;
+        var dist = 0f;
+        for(CoreBuild core : team.cores()){
+            if(result == null || core.block.canReplace(result.block) || (result.block == core.block && core.dst(player.x, player.y) < dist)){
```
Comment (Quezler) [inline core/src/mindustry/entities/comp/PlayerComp.java:62]:
```suggestion
            if(result == null || core.block.size >result.block.size || (result.block == core.block && core.dst(player.x, player.y) < dist)){
```

File: core/src/mindustry/entities/comp/PlayerComp.java [hunk 3]
```diff
@@ -124,19 +134,21 @@ public void update(){
...
-        }else if(core != null){
-            //have a small delay before death to prevent the camera from jumping around too quickly
-            //(this is not for balance, it just looks better this way)
-            deathTimer += Time.delta;
-            if(deathTimer >= deathDelay){
-                //request spawn - this happens serverside only
-                core.requestSpawn(self());
-                deathTimer = 0;
+        }else{
+            CoreBuild core = closestBestCore();
```
Comment (Quezler) [inline core/src/mindustry/entities/comp/PlayerComp.java:138]:
```suggestion
            CoreBuild core = respawnCore();
```

**Review by Quezler — CHANGES_REQUESTED** [review]
on the debate of `best/worst/largestNearbyCore/closestLargestCore/etc` i would propose to just name the function `respawnCore`, sounds way calmer and only exists on the player object 🤔

**Review by Quezler — APPROVED** [review]

File: core/src/mindustry/entities/comp/PlayerComp.java [hunk 4]
```diff
@@ -55,6 +55,18 @@ public boolean isBuilder(){
         return state.teams.closestCore(x, y, team);
     }
 
+    public @Nullable CoreBuild respawnCore(){
+        CoreBuild result = null;
+        var dist = 0f;
+        for(CoreBuild core : team.cores()){
+            if(result == null || core.block.size > result.block.size || (result.block.size == core.block.size && core.dst(player.x, player.y) < dist)){
```
Comment (Quezler) [inline core/src/mindustry/entities/comp/PlayerComp.java:62]:
```suggestion
            if(result == null || core.block.size > result.block.size || (core.block.size == result.block.size && core.dst(x, y) < dist)){
```

File: core/src/mindustry/entities/comp/PlayerComp.java [hunk 5]
```diff
@@ -55,6 +55,18 @@ public boolean isBuilder(){
...
+                result = core;
+                dist = core.dst(player.x, player.y);
```
Comment (Quezler) [inline core/src/mindustry/entities/comp/PlayerComp.java:64]:
```suggestion
                dist = core.dst(x, y);
```

**Review by Quezler — APPROVED** [review]

**Review by summetdev — APPROVED** [review]

File: core/src/mindustry/entities/comp/PlayerComp.java [hunk 6]
```diff
@@ -55,6 +55,18 @@ public boolean isBuilder(){
         return state.teams.closestCore(x, y, team);
     }
 
+    public @Nullable CoreBuild respawnCore(){
+        CoreBuild result = null;
+        var dist = 0f;
+        for(CoreBuild core : team.cores()){
+            if(result == null || core.block.size > result.block.size || (result.block.size == core.block.size && dst(core) < dist)){
```
Comment (joshuaptfan) [inline core/src/mindustry/entities/comp/PlayerComp.java:62]:
```suggestion
            if(result == null || result.block.size < core.block.size || (result.block.size == core.block.size && dst(core) < dist)){
```
Sorry, OCD

**Review by joshuaptfan — CHANGES_REQUESTED** [review]

**Review by Voz-Duh — APPROVED** [review]

**Review by Quezler — APPROVED** [review]

**Comment by Anuken** [comment]
Implemented manually.

**Pull request closed without merging**